mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala

```diff
@@ -0,0 +1,346 @@
+ * Model fitted by FPGrowth.
+ *
+ * @param freqItemsets frequent items in the format of DataFrame("items"[Seq], "freq"[Long])
+ */
+@Since("2.2.0")
+@Experimental
+class FPGrowthModel private[ml] (
+    @Since("2.2.0") override val uid: String,
+    @transient val freqItemsets: DataFrame)
+  extends Model[FPGrowthModel] with FPGrowthParams with MLWritable {
+
+  /** @group setParam */
+  @Since("2.2.0")
+  def setMinConfidence(value: Double): this.type = set(minConfidence, value)
+
+  /** @group setParam */
+  @Since("2.2.0")
+  def setFeaturesCol(value: String): this.type = set(featuresCol, value)
+
+  /** @group setParam */
+  @Since("2.2.0")
+  def setPredictionCol(value: String): this.type = set(predictionCol, value)
+
+  /**
+   * Get association rules fitted by AssociationRules using the minConfidence. Returns a dataframe
+   * with three fields, "antecedent", "consequent" and "confidence", where "antecedent" and
+   * "consequent" are Array[T] and "confidence" is Double.
+   */
+  @Since("2.2.0")
+  @transient lazy val associationRules: DataFrame = {
+    val freqItems = freqItemsets
+    AssociationRules.getAssociationRulesFromFP(freqItems, "items", "freq", $(minConfidence))
+  }
+
+  /**
+   * The transform method first generates the association rules according to the frequent itemsets.
+   * Then for each association rule, it will examine the input items against antecedents and
+   * summarize the consequents as prediction. The prediction column has the same data type as the
+   * input column. (Array[T])
+   * Note that internally it uses Cartesian join and may exhaust memory for large datasets.
+   */
+  @Since("2.2.0")
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    transformSchema(dataset.schema, logging = true)
+    genericTransform(dataset)
+  }
+
+  private def genericTransform[T](dataset: Dataset[_]): DataFrame = {
+    // use index to perform the join and aggregateByKey, and keep the original order after join.
+    val indexToItems = dataset.select($(featuresCol)).rdd.map(r => r.getSeq[T](0))
+      .zipWithIndex().map(_.swap)
+    val rulesRDD = associationRules.select("antecedent", "consequent").rdd
+      .map(r => (r.getSeq[T](0), r.getSeq[T](1)))
+
+    val indexToConsequents = indexToItems.cartesian(rulesRDD).map {
+      case ((id, items), (antecedent, consequent)) =>
+        val consequents = if (items != null) {
+          val itemSet = items.toSet
+          if (antecedent.forall(itemSet.contains)) {
+            consequent.filterNot(itemSet.contains)
+          } else {
+            Seq.empty
+          }
+        } else {
+          Seq.empty
+        }
+        (id, consequents)
+    }.aggregateByKey(new ArrayBuffer[T])((ar, seq) => ar ++= seq, (ar, seq) => ar ++= seq)
+     .map { case (index, cons) => (index, cons.distinct) }
+
+    val rowAndConsequents = dataset.toDF().rdd.zipWithIndex().map(_.swap)
+      .join(indexToConsequents).sortByKey(ascending = true, dataset.rdd.getNumPartitions)
```

Comment:
If that's the case, does zipWithIndex() guarantees the same result for multiple calls? 
And if can ignore the ordering, perhaps I have a faster way to write the transform.